### PR TITLE
util/stream: Add pumping functions

### PR DESCRIPTION
**Pull request by kasjer**
This extends in_stream and out_stream to provide pumping
functions

Additionally stream_pump() function is added to allow easy
pumping data between streams.

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

### CODE REVIEW

File: hw/drivers/i2s/src/i2s_stream.c [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    int written = 0;
+    uint32_t available = istream_available(istream);
+    uint32_t sample_count;
+
+    if (available < count) {
+        count = available;
+    }
+    sample_count = count / i2s->sample_size_in_bytes;
+
+    while (count) {
```
Comment (sjanc) [inline hw/drivers/i2s/src/i2s_stream.c:44]: this looks bit odd since count is never modified inside loop

File: util/stream/src/stream.c [hunk 2]
```diff
@@ -107,3 +108,30 @@ istream_read(struct in_stream *istream, uint8_t *buf, uint32_t count)
...
+    if (ostream->vft->pump_from) {
+        pumped = ostream->vft->pump_from(ostream, istream, count);
+    } else if (istream->vft->pump_to) {
+        /* When input stream has pump_to used it */
+        pumped = istream->vft->pump_to(istream, ostream, count);
+    } else {
+        /* Both stream don't provide pump functions, use local buffer for pumping data */
+        uint8_t buf[16];
+        pumped = 0;
+        while (count) {
```
Comment (sjanc) [inline util/stream/src/stream.c:126]: same here

Comment (kasjer) [reply util/stream/src/stream.c:126]: fixed

File: hw/drivers/i2s/src/i2s_stream.c (hunk 1 above)
Comment (kasjer) [reply hw/drivers/i2s/src/i2s_stream.c]: loop was not needed and now is removed it could be replaced with if

**Review by sjanc — APPROVED** [review]

**Pull request merged**